### DISCUSSION TITLE
how to update some obj field value when use select

### DISCUSSION THREAD

**Discussion by flymzero** [rrousselGit/riverpod#1604] [Q&A] [answered]
i want know what is the right way to change a field value(when use select).

if is **ref.read(provider.state).state =  xxx** , if obj have many field then each field needs to be assigned a value.

```dart
/// this is a object
class Obj {
  int a;
  int b;

 Obj(this.a, this.b);
}

final provider = StateProvider<Obj>((ref) {
  return Obj(1,1);
});

/// in some widget 
@override
  Widget build(BuildContext context, WidgetRef ref) {
  final a = ref.watch(provider.select((v) => v.a));
  return Text('$a');
}

/// i want to change the obj field a value and can watch
/// in a button on onPressed function
...
onPressed: () {
  // it is not work!!!
  ref.read(provider).a = 2;
  // it is work
  ref.read(provider.state).state = Obj(2,1);
}
...

```

**Comment by Clashkid155** [comment]
You should try
```dart
ref.read(provider.notifier).update((state) => state.a = 2);
```
You can read more at [Riverpod](https://riverpod.dev/docs/providers/state_provider/#how-to-update-the-state-based-on-the-previous-value-without-reading-the-provider-twice)

**Reply by flymzero** [reply]
thanks, but it is also not work

**Comment by rrousselGit** [comment, marked as answer]
That's expected.
The state of a StateProvider is supposed to be immutable, yet you used a mutable class.


You are supposed to clone the object, like you did in your second snippet.
